src/typechecker/FStar.TypeChecker.Rel.fs

```diff
@@ -3200,113 +3198,140 @@ and solve_c (env:Env.env) (problem:problem<comp>) (wl:worklist) : solution =
+                stronger_t_shape_error "not an arrow or not enough binders") r in
+
+          let rest_bs_uvars, g_uvars = Env.uvars_for_binders env rest_bs
+            [NT (a_b |> fst, c2.result_typ)]
+            (fun b -> BU.format3 "implicit for binder %s in subcomp of %s at %s"
+              (Print.binder_to_string b) (Ident.string_of_lid c2.effect_name) (Range.string_of_range r)) r in
+
+          let wl = { wl with wl_implicits = g_uvars.implicits@wl.wl_implicits } in  //AR: TODO: FIXME: using knowledge that g_uvars is only implicits
+
+          let substs = List.map2
+            (fun b t -> NT (b |> fst, t))
+            (a_b::rest_bs) (c2.result_typ::rest_bs_uvars) in
+
+          let f_sub_probs, wl =
+            let f_sort_is = U.effect_indices_from_repr
+              (f_b |> fst).sort
+              (Env.is_layered_effect env c1.effect_name)
+              r (stronger_t_shape_error "type of f is not a repr type")
+              |> List.map (SS.subst substs) in
+
+            List.fold_left2 (fun (ps, wl) f_sort_i c1_i ->
+              if Env.debug env <| Options.Other "LayeredEffectsEqns"
+              then BU.print3 "Layered Effects (%s) %s = %s\n" subcomp_name
+                     (Print.term_to_string f_sort_i) (Print.term_to_string c1_i);
+              let p, wl = sub_prob wl f_sort_i EQ c1_i "indices of c1" in
+              ps@[p], wl
+            ) ([], wl) f_sort_is (c1.effect_args |> List.map fst) in
+
+          let stronger_ct = stronger_c |> SS.subst_comp substs |> U.comp_to_comp_typ in
+
+          let g_sub_probs, wl =
+            let g_sort_is = U.effect_indices_from_repr
+              stronger_ct.result_typ
+              (Env.is_layered_effect env c2.effect_name)
+              r (stronger_t_shape_error "subcomp return type is not a repr") in
+
+            List.fold_left2 (fun (ps, wl) g_sort_i c2_i ->
+              if Env.debug env <| Options.Other "LayeredEffectsEqns"
+              then BU.print3 "Layered Effects (%s) %s = %s\n" subcomp_name
+                     (Print.term_to_string g_sort_i) (Print.term_to_string c2_i);
+              let p, wl = sub_prob wl g_sort_i EQ c2_i "indices of c2" in
+              ps@[p], wl
+            ) ([], wl) g_sort_is (c2.effect_args |> List.map fst) in
+
+          let fml =
+            let u, wp = List.hd stronger_ct.comp_univs, fst (List.hd stronger_ct.effect_args) in
+            Env.pure_precondition_for_trivial_post env u stronger_ct.result_typ wp Range.dummyRange in
+
+          let sub_probs = ret_sub_prob::(is_sub_probs@f_sub_probs@g_sub_probs@(g_lift.deferred |> List.map snd)) in
+          let guard =
+            let guard = U.mk_conj_l (List.map p_guard sub_probs) in
+            match g_lift.guard_f with
+            | Trivial -> guard
+            | NonTrivial f -> U.mk_conj guard f in
+          let wl = solve_prob orig (Some <| U.mk_conj guard fml) [] wl in
+          solve env (attempt sub_probs wl) in
```

Comment:
This is a big diff. Can you summarize the main intent of the change here?

Reply:
Not much actually. This code is tweaked a little to support `polymonadic_subcomp`. Earlier to solve a `c1 <: c2`, we would first lift `c1` to `c2` or vice versa, and then call either `solve_sub` or `solve_layered_sub` depending on whether the resultant effect is layered.

But now, such a lift may not exist, and instead there may be a polymoandic subcomp. So, for `c1 <: c2`, this code says if either is a layered effect go to `solve_layered_sub` which will decide whether to lift or use a polymonadic subcomp.